cpp/src/arrow/python/pandas_convert.cc

```diff
@@ -419,21 +420,45 @@ inline Status PandasConverter::ConvertData<BooleanType>(std::shared_ptr<Buffer>*
 
 Status InvalidConversion(PyObject* obj, const std::string& expected_type_name) {
   PyObject* type = PyObject_Type(obj);
-  RETURN_IF_PYERROR();
+
+  if (type == nullptr) {
+    DCHECK_NE(PyErr_Occurred(), nullptr)
+        << "Call to PyObject_Type(...) returned NULL but PyErr_Occurred() was NULL";
+    RETURN_IF_PYERROR();
+  }
 
   PyObject* type_name = PyObject_GetAttrString(type, "__name__");
-  if (PyErr_Occurred()) { Py_DECREF(type); }
-  RETURN_IF_PYERROR();
 
-  Py_ssize_t size;
-  const char* bytes = PyUnicode_AsUTF8AndSize(type_name, &size);
+  if (type_name == nullptr) {
+    Py_DECREF(type);
+    DCHECK_NE(PyErr_Occurred(), nullptr) << "Call to PyObject_GetAttrString(...) "
+                                            "returned NULL but PyErr_Occurred() was NULL";
+    RETURN_IF_PYERROR();
+  }
+
+  PyObject* bytes_obj = PyUnicode_AsUTF8String(type_name);
+
+  if (bytes_obj == nullptr) {
+    Py_DECREF(type_name);
+    Py_DECREF(type);
+    DCHECK_NE(PyErr_Occurred(), nullptr) << "Call to PyObject_AsUTF8String(...) returned "
+                                            "NULL but PyErr_Occurred() was NULL";
```

Comment:
Yikes. Perhaps we should develop some tools to help with the Python C API in C++. I have tried to use the OwnedRef to handle RAII-based DECREF